File: client/cocos/assets/scripts/frame/reddot/RedDotNode.ts
import { _decorator } from 'cc';
import { RedDotComponent } from './RedDotCompoent';
const { ccclass, property } = _decorator;

@ccclass('RedDotNode')
export class RedDotNode {
  /**
   * 分数
   */
  private _score: number = 0;

  private _parent: RedDotNode;

  private _children: Map<string, RedDotNode> = new Map<string, RedDotNode>();

  private _ui: RedDotComponent = null;

  public addChild(name: string): RedDotNode {
    if (this._children.has(name)) {
      return this._children.get(name);
    }
    const node = new RedDotNode();
    this._children.set(name, node);
    node._parent = this;
    return node;
  }

  public set ui(ui: RedDotComponent) {
    this._ui = ui;
  }

  public get ui(): RedDotComponent {
    return this._ui;
  }

  public set score(score: number) {
    this._score = score;
  }

  public get score(): number {
    return this._score;
  }

  public get parent(): RedDotNode {
    return this._parent;
  }

  public get children(): Map<string, RedDotNode> {
    return this._children;
  }
}

File: client/unity/Assets/Scripts/Frame/Commons/RedDot/RedDotNode.cs
﻿namespace Nova.Commons.RedDot
{
    using System.Collections.Generic;
    using UnityEngine;

    /// <summary>
    /// 红点节点
    /// </summary>
    public class RedDotNode  
    {
       
       public int score {
           get; set;
       }


        public RedDotComponent ui  {
           get;  set;
        }


        /// <summary>
        /// 父节点
        /// </summary>
       public RedDotNode parent{
        get  ; set;
       }

          /// <summary>
          /// 构造函数
          /// </summary>
          /// <param name="score">初始分数</param>
        public RedDotNode(int score = 0) {
            this.score = score;
            this.children = new Dictionary<string, RedDotNode>();
        }
       /// <summary>
       /// 子节点
       /// </summary>
       public Dictionary<string, RedDotNode> children  {
        get; private set;
       }
        
        public Dictionary<string, RedDotNode> GetChildren()
        {
            return this.children;
        }
        public RedDotNode AddChild(string key)  {
           if (this.children.TryGetValue(key, out var child)) {
               return child;
           }
           child = new RedDotNode();
           child.parent = this;
           this.children[key] = child;
           return child;
        }
    }
}
